client/allocrunner/groupservice_hook.go

```diff
@@ -117,10 +128,30 @@ func (h *groupServiceHook) Update(req *interfaces.RunnerUpdateRequest) error {
 	return h.consulClient.UpdateWorkload(oldWorkloadServices, newWorkloadServices)
 }
 
-func (h *groupServiceHook) Postrun() error {
+func (h *groupServiceHook) PreKill() {
 	h.mu.Lock()
 	defer h.mu.Unlock()
+
+	// If we have a shutdown delay deregister
+	// group services and then wait
+	// before continuing to kill tasks
 	h.deregister()
+	h.deregistered = true
+
+	h.logger.Debug("waiting before removing group service", "shutdown_delay", h.delay)
```

Comment:
Let's shortcircuit before the log line if `h.delay == 0`. Would hate to spew confusing logs in the default case even if its debug only.